lib/Dialect/HW/HWOps.cpp

```diff
@@ -1785,6 +1785,56 @@ LogicalResult ArrayCreateOp::verify() {
   return success();
 }
 
+// Check whether an integer value is an offset from a base.
+static bool isOffset(Value base, Value index, uint64_t offset) {
+  if (auto constBase = base.getDefiningOp<hw::ConstantOp>())
+    if (auto constIndex = index.getDefiningOp<hw::ConstantOp>())
+      return constBase.getValue() + offset == constIndex.getValue();
+  return false;
+}
+
+// Canonicalize a create of consecutive elements to a slice.
+static LogicalResult foldCreateToSlice(ArrayCreateOp op,
+                                       PatternRewriter &rewriter) {
+  // Do not canonicalize create of get into a slice.
+  auto arrayTy = hw::type_cast<ArrayType>(op.getType());
+  if (arrayTy.getSize() <= 1)
+    return failure();
+
+  // Check if create arguments are consecutive elements of the same array.
+  Value input;
+  Value lowIndex;
+  for (const auto &[i, v] : llvm::enumerate(llvm::reverse(op.getInputs()))) {
+    auto get = v.getDefiningOp<ArrayGetOp>();
+    if (!get || (input && get.getInput() != input))
+      return failure();
+    input = get.getInput();
+    if (!lowIndex) {
+      lowIndex = get.getIndex();
+    } else {
+      if (!isOffset(lowIndex, get.getIndex(), i))
+        return failure();
+    }
+  }
+
+  // If the whole array is sliced, return it in full.
+  if (input.getType() == op.getType()) {
+    rewriter.replaceOp(op, input);
```

Comment:
Is there a regression test for this branch somewhere?

Reply:
I've added a test now.